Autoscaler Rancher Provider Token Privileges Issue

<!--
Please answer these questions before submitting your bug report. Thanks!
-->

**Which component are you using?**:
Autoscaler with rancher provider.

**What version of the component are you using?**:
helm chart version:
sources:
  - https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler
type: application
version: 9.28.0

rancher provider version:
image:
  repository: registry.k8s.io/autoscaling/cluster-autoscaler
  tag: v1.26.1

Rancher version: 2.6.11

Component version:

**What k8s version are you using (`kubectl version`)?**:
kubectl version --short
Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
Client Version: v1.26.2
Kustomize Version: v4.5.7
Server Version: v1.24.10+rke2r1
WARNING: version difference between client (1.26) and server (1.24) exceeds the supported minor version skew of +/-1

**What environment is this in?**:
I am using rancher to deploy clusters via plugins (aws or azure).
For each cluster, i am deploying autoscaler for cluster autoscaling.
I need to limit the cluster autoscaler privileges to its cluster only.

**What did you expect to happen?**:
Autoscaler should do the scale up/down by using rancher token. I want to limit the this token privileges to a certain cluster. I created a user (standard user) and made it cluster owner with no other global role/privileges given.

**What happened instead?**:
the auto scaler pod crashes after 30s and it throws error as given below. It works only if i add a global role with list and update privileges for certain rancher APIs. It made the token capable to send update calls for all clusters managed by rancher.


**How to reproduce it (as minimally and precisely as possible)**:
1. Create rancher user (standard user) and make it cluster owner with no other global role/privileges set.
2. Deploy autoscaler chart on your kubernetes. 
3. Note: there are some typos in the helm chart deployment file that needs fixing. Some parts of the template file does not have relevant if condition for rancher provider. I will create a separate issue for this.
4. Use rancher provider image tag v1.26.1 
5. The pod will fail after 30s roughly.
6. Check the pod logs to see the error pasted below.



**Anything else we need to know?**:
Here is the error that appears in the pod logs:

I0405 08:12:31.455800       1 rancher_provider.go:234] scalable node group found: node1 (3:5)
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x391ff36]

goroutine 77 [running]:
k8s.io/autoscaler/cluster-autoscaler/cloudprovider/rancher.(*nodeGroup).listMachines(0xc00038f8f0)
        /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/rancher/rancher_nodegroup.go:348 +0x236
k8s.io/autoscaler/cluster-autoscaler/cloudprovider/rancher.(*nodeGroup).machineByName(0xc00038f8f0, {0xc000e2aea0, 0x26})
        /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/rancher/rancher_nodegroup.go:353 +0x45
k8s.io/autoscaler/cluster-autoscaler/cloudprovider/rancher.(*RancherCloudProvider).NodeGroupForNode(0xc001010380, 0xc000fe4000)
        /gopath/src/k8s.io/autoscaler/cluster-autoscaler/cloudprovider/rancher/rancher_provider.go:152 +0x1b7
k8s.io/autoscaler/cluster-autoscaler/processors/nodeinfosprovider.(*MixedTemplateNodeInfoProvider).Process.func1(0xc000fe4000?)
        /gopath/src/k8s.io/autoscaler/cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor.go:86 +0x56
k8s.io/autoscaler/cluster-autoscaler/processors/nodeinfosprovider.(*MixedTemplateNodeInfoProvider).Process(0xc000c92cc0, 0xc000f6cc00, {0xc000ac5c60, 0x4, 0x1?}, {0xc000ac5cc0, 0x3, 0x4}, 0xc000cb5380, {0xc103693bdaa52e01, ...})
        /gopath/src/k8s.io/autoscaler/cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor.go:115 +0x22d
k8s.io/autoscaler/cluster-autoscaler/core.(*StaticAutoscaler).RunOnce(0xc00087ff40, {0x4?, 0xc000179710?, 0x87e1e20?})
        /gopath/src/k8s.io/autoscaler/cluster-autoscaler/core/static_autoscaler.go:335 +0xba3
main.run(0xc000b02180?, {0x5a277b8, 0xc00096ba40})
        /gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:449 +0x2cd
main.main.func2({0x8b9dc6?, 0xc000b02180?})
        /gopath/src/k8s.io/autoscaler/cluster-autoscaler/main.go:536 +0x25
created by k8s.io/client-go/tools/leaderelection.(*LeaderElector).Run
        /gopath/src/k8s.io/autoscaler/cluster-autoscaler/vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:211 +0x11b

--- a/cluster-autoscaler/cloudprovider/rancher/rancher_nodegroup.go
+++ b/cluster-autoscaler/cloudprovider/rancher/rancher_nodegroup.go
@@ -344,9 +344,12 @@ func (ng *nodeGroup) listMachines() ([]unstructured.Unstructured, error) {
 			LabelSelector: fmt.Sprintf("%s=%s-%s", machineDeploymentNameLabelKey, ng.provider.config.ClusterName, ng.name),
 		},
 	)
+	if err != nil {
+		return nil, fmt.Errorf("could not list machines: %w", err)
+	}
 
 	ng.machines = machinesList.Items
-	return machinesList.Items, err
+	return machinesList.Items, nil
 }
 
 func (ng *nodeGroup) machineByName(name string) (*unstructured.Unstructured, error) {

